src/avro/src/schema.rs

```diff
@@ -579,18 +580,52 @@ impl fmt::Debug for FullName {
 pub type Documentation = Option<String>;
 
 impl Name {
-    /// Create a new `Name`.
-    /// No `namespace` nor `aliases` will be defined.
-    pub fn new(name: &str) -> Name {
-        Name {
-            name: name.to_owned(),
-            namespace: None,
-            aliases: None,
+    /// Reports whether the given string is a valid Avro name.
+    pub fn is_valid(name: &str) -> bool {
+        static MATCHER: Lazy<Regex> =
+            Lazy::new(|| Regex::new(r"(^[A-Za-z_][A-Za-z0-9_]*)$").unwrap());
+        MATCHER.is_match(name)
+    }
+
+    /// Returns an error if the given name is invalid.
+    pub fn validate(name: &str) -> Result<(), AvroError> {
+        match Self::is_valid(name) {
+            true => Ok(()),
+            false => {
+                Err(ParseSchemaError::new(format!(
+                    "Invalid name. Must start with [A-Za-z_] and subsequently only contain [A-Za-z0-9_]. Found: {}",
+                    name
+                )).into())
+            }
+        }
+    }
+
+    /// Rewrites `name` to be valid.
+    ///
+    /// Any non alphanumeric characters are replaced with underscores. If the
+    /// name begins with a number, it is prefixed with `_`.
+    pub fn make_valid(name: &str) -> String {
+        let mut out = String::new();
+        let mut chars = name.chars();
+        match chars.next() {
+            Some(ch @ ('a'..='z' | 'A'..='Z')) => out.push(ch),
+            Some(ch @ '0'..='9') => {
+                out.push('_');
+                out.push(ch);
+            }
+            _ => out.push('_'),
+        }
+        for ch in chars {
+            match ch {
+                '0'..='9' | 'a'..='z' | 'A'..='Z' => out.push(ch),
+                _ => out.push('_'),
```

Comment:
we assert later on that these names are still unique right?

Reply:
Er, we don't assert, but we restore the uniqueness by appending `1`, `2`, and so on.